### PR TITLE
The great BI merge of 2016

### PR DESCRIPTION
**Pull request by ewels**
Ok, I'm gearing up to release v0.4 of Cluster Flow, this v0.4dev madness has gone on for long enough. So, time to try to bring in all of your new code.

The PR below looks relatively sensible, main problem is the new modules that still use the old syntax for requesting resources. They won't work with the current CF code. I can update them so that they do, but then I'd really like you guys to start using the newly released v0.4 version so we don't have to do this again ;)

Sounds doable?

### CODE REVIEW

**Comment by ewels** [comment]
ps. Other pet hate is leaving commented out code in the package, so I'll probably delete that too if that's ok.

**Comment by ewels** [comment]
Ok, so proposed workflow:

1. I merge this PR into a new branch on _my_ repository
2. I do my own updates there to make the code compatible with current version of cluster flow
3. I do a branch PR back into master and you guys can look over what I've done
4. I release v0.4
5. You guys merge new v0.4 code into your repo
6. You guys actually use v0.4 :)

Sound ok @s-andrews @FelixKrueger?

File: source/CF/Helpers.pm [hunk 1]
```diff
@@ -275,7 +275,7 @@ sub is_bam_paired_end {
 	my $header_found;
 	foreach (split("\n", $headers)){
 		if(/^\@PG/){
-			if(/-1/ && /-2/){
+			if(/\s+-1\s+/ && /\s+-2\s+/){
```
Comment (ewels) [inline source/CF/Helpers.pm]: This entire function has been replaced in my current version:

```perl
sub is_bam_paired_end {

	# Load samtools
	my @modules = ('samtools');
	my %loaded_mods = ();
	&CF::Helpers::load_environment_modules(\@modules,\%loaded_mods);

	my ($file) = @_;

	unless($file =~ /.bam$/ || $file =~ /.sam$/){
		warn "\n$file is not a .bam or .sam file - can't figure out PE / SE mode..\nExiting..\n\n";
		die;
	}

    # Read the first 1000 lines withx samtools
    my $se_reads = 0;
    my $pe_reads = 0;
    my $readcount = 0;
    open(my $fh, '-|', "samtools view $file") or die "Could not run samtools to check BAM pe/se: $!";
    while (<$fh>) {
        last if ($readcount >= 1000);
        my ($flag) = (split (/\t/))[1];
        if ($flag & 0x1){
            $pe_reads++;
        } else {
            $se_reads++;
        }
        $readcount++;
    }
    close($fh);

    # Look at our counts
    if($pe_reads >= 800){
        return 1;
    } else {
        return 0;
    }

}
```

Any objections to me ignoring your change and using my version of the function?

Comment (FelixKrueger) [inline source/CF/Helpers.pm]: As long as it works fine that's absolutely fine.

**Comment by s-andrews** [comment]
Sure - works for me.  Would be good to try the new repo as soon as you have
it merged - I'd still like to get some of the file naming issues sorted out
before a final 0.4 release so we don't end up working off a branch again.

Simon.

From:  Phil Ewels <notifications@github.com>
Reply-To:  ewels/clusterflow
<reply+007c7ae36dbba92039c362bf1c949d76c5d30dc47b085bbe92cf000000011372667f9
2a169ce098386ee@reply.github.com>
Date:  Friday, 10 June 2016 at 12:29
To:  ewels/clusterflow <clusterflow@noreply.github.com>
Cc:  Simon Andrews <simon@proteo.me.uk>, Mention
<mention@noreply.github.com>
Subject:  Re: [ewels/clusterflow] The great BI merge of 2016 (#88)


Ok, so proposed workflow:
1. I merge this PR into a new branch on my repository
2. I do my own updates there to make the code compatible with current
version of cluster flow
3. I do a branch PR back into master and you guys can look over what I've
done
4. I release v0.4
5. You guys merge new v0.4 code into your repo
6. You guys actually use v0.4 :)
Sound ok @s-andrews <https://github.com/s-andrews>  @FelixKrueger
<https://github.com/FelixKrueger> ?


You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/ewels/clusterflow/pull/88#issuecomment-225158750> , or
mute the thread 
<https://github.com/notifications/unsubscribe/AHx640eAOsruBdOPWj2JNxkPKhKNv8
3zks5qKUp_gaJpZM4Iy3Jy> .

**Comment by ewels** [comment]
Ok, I finished sorting out all of the merge conflicts. See commit 82afc086014c3d9269488f53e4d1dc264403bfb4

I admit I thought that this would then class this PR as merged, but apparently not. Maybe I have to merge your commits into master for that.

Anyway, I will do some more updating on that branch and then merge into master when I'm happy.

**Pull request merged**